Move tech design overview from wiki into repo (#12601)
nvaccess/nvda: devDocs/technicalDesignOverview.md

An app module provides support specific to an application for these cases.
An app module is derived from the `appModuleHandler.AppModule` base class.
App modules receive events for all [NVDA objects](#nvda-objects) in the application and can bind scripts which can be executed anywhere in that application.
They can also implement their own NVDA objects for use within the application.

#### Global Plugins
Aside from application specific customisation using [app modules](#app-modules), it is also possible to extend NVDA on a global level.
For example, new global commands can be added, behaviour can be changed and new GUI toolkits can be supported.
This can be done using global plugins.
A global plugin is derived from the `globalPluginHandler.GlobalPlugin` base class.
Similar to [global commands](#global-commands), they can bind scripts which can be executed everywhere.
They can also implement their own global [NVDA Objects](#nvda-objects).

### Tree Interceptors
Sometimes, it is necessary to intercept events and scripts for an entire hierarchy (or tree) of [NVDA objects](#nvda-objects).

> For example, this is necessary to seemlessly handle complex documents which consist of many objects.

Reviewer: seemlessly -> seamlessly

Each API provides a different way of querying and manipulating text.
Just as NVDA objects provide an abstract representation of a widget, TextInfo objects provide an abstract representation of a range of text.
These objects are derived from the `textInfos.TextInfo` base class.

TextInfo objects contain properties and methods to:
* Move or expand the range by units such as character, word, line and paragraph
* compare the start and end of a range with itself or another range
* Fetch the text and formatting of the range
 
 You can fetch a TextInfo object from an NVDA object via its `makeTextInfo` method, passing in the particular `textInfos.POSITION_*` constant depending on whether you want to fetch a range representing the position of the caret, selection, start or end of the text, or the entire text.
 
### Global Commands
The global commands object (`globalCommands.GlobalCommands`) contains built-in global scripts; i.e.
they can be executed everywhere.
For example, the review, report current focus and date/time scripts are all located in global commands.

### Plugins

> NvDA allows third-parties to extend NvDA's functionality through plugins and add-ons.

Reviewer 2: ```suggestion
NVDA allows third-parties to extend NVDA's functionality through plugins and add-ons.
```

An app module provides support specific to an application for these cases.
An app module is derived from the `appModuleHandler.AppModule` base class.
App modules receive events for all [NVDA objects](#nvda-objects) in the application and can bind scripts which can be executed anywhere in that application.
They can also implement their own NVDA objects for use within the application.

#### Global Plugins
Aside from application specific customisation using [app modules](#app-modules), it is also possible to extend NVDA on a global level.
For example, new global commands can be added, behaviour can be changed and new GUI toolkits can be supported.
This can be done using global plugins.
A global plugin is derived from the `globalPluginHandler.GlobalPlugin` base class.
Similar to [global commands](#global-commands), they can bind scripts which can be executed everywhere.
More specifically, global plugins receive events for all [NVDA objects](#nvda-objects) in the Operating System and can bind scripts which can be executed anywhere.
They can also implement their own global [NVDA Objects](#nvda-objects).

### Tree Interceptors

> Sometimes, it is necessary to intercept events and scripts for an entire hierarchy (or tree) of [NVDA objects](#nvda-objects).

Reviewer 3: This leaves me wondering, intercept between which two points? Where is the event "coming from", and where is it "going"?

For example, this is necessary to seamlessly handle complex documents which consist of many objects.
This can be done using a tree interceptor.
A tree interceptor (TreeInterceptor) is derived from the `treeInterceptorHandler.TreeInterceptor` base class.
It receives events and scripts for all [NVDA objects](#nvda-objects) beneath and including the root NVDA object of the tree interceptor.
Tree interceptors are created when a TreeInterceptor class is returned from the `treeInterceptorClass` property of an NVDA object.
Tree interceptors are used mostly for web documents, where all events and scripts for NVDA objects within a document need to be handled by the document (root NVDA object) itself.

#### Browse mode documents
Complex documents such as web pages are very often not flat; i.e. information does not simply run from top to bottom.
Because of this, complex document browsers often do not provide a way to navigate documents using the caret, and even when they do, it is often problematic.
Therefore, screen readers need to create their own flat representation of a document from the object hierarchy provided by the browser and allow the user to navigate this flat representation.

> Browse mode documents are a subclass of `TreeInterceptor` that provide scripts that allow navigating the document in a linea fassion.

Reviewer 2: ```suggestion
Browse mode documents are a subclass of `TreeInterceptor` that provide scripts that allow navigating the document in a linear fashion.
```

## Special Object Functions

### Events
NVDA object, global plugin, app module and tree interceptor instances can all contain special methods which handle events for NVDA Objects.
These methods are all named beginning with "event_"; e.g. `event_gainFocus` and `event_nameChange`.
These events are generally executed by a call to `eventHandler.executeEvent`, which is in turn generally called resultant to events queued by [API Handlers](#api-handlers).
Most events do not take any additional arguments.
Global plugins, app modules and tree interceptors are passed a handler function which should be called if the event should be handled by the next handler;
e.g. the object itself.

Although an event is always for a particular NVDA object, it first has a chance of being handled by global plugins, app modules or tree interceptors.
If an event is handled by one of these, meaning that an `event_*` method was found and executed, the event stops there and does not go further, unless the method that handled it specifically calls the `nextHandler` function object passed to it. 

The chain of handlers is as follows:
* The first found global plugin

> * The next found global plugin...

Reviewer 2: ```suggestion
* The next found global plugin (until no more are found)
```

* The tree interceptor containing the NVDA object the event is for. I.e. fetched from the NVDA object's `treeInterceptor` property if the property is not `None`
* The NVDA object itself.

### Scripts
NVDA object, global plugin, app module and tree interceptor instances can all contain special methods called scripts which are executed in response to [input gestures](#input-gestures) from the user.
These methods are all named beginning with "script_"; e.g. `script_reportCurrentFocus` and `script_dateTime`.
Script methods are passed the input gesture that triggered them.
Input gestures are bound to scripts in the class using a `scriptHandler.script` function decorator.
They can also be bound at runtime using `bindGesture`.
These are inherited from `baseObject.ScriptableObject`.

Similar to events, input gestures have a chance to be handled by a script at one of many levels.
But unlike events, once an input gesture finds and executes a script, there is no clean way to have the input gesture handled by a subsequent level.
 
 The chain of handlers is as follows:
* The first found global plugin 

> * The next found global plugin...

Reviewer 2: ```suggestion
* The next found global plugin (until no more are found)
```

* The app module containing the currently focused NVDA object, I.e. fetched from the NVDA object's `appModule` property 
* The tree interceptor containing the currently focused NVDA object, I.e. fetched from the NVDA object's `treeInterceptor` property if the property is not `None`
* The currently focused NVDA object
* the first ancestor (parent) of the currently focused NVDAObject, if the found script's `canPropagate` property is True
* the next ancestor of the currently focused NVDAObject, if the found script's `canPropagate` property is True...
 * Global commands
 
## Inter-process Communication
In general terms, every running application or service on a computer, including NVDA, is a separate process.
No process can access data in another process except via special operating system mechanisms.

> This is called inter-process communication (IPC).

Reviewer 3: It would be helpful if the specific IPC approaches used by NVDA were described.
